tests/buffer/buffer_data_tracking.cpp

```diff
@@ -0,0 +1,160 @@
+/* RUN: %{execute}%s | %{filecheck} %s
+   CHECK: Result:
+   CHECK-NEXT: 0 0 0
+   CHECK-NEXT: 3 4 5
+   CHECK-NEXT: 1 1 1
+*/
+#include <CL/sycl.hpp>
+#include <iostream>
+#include "test-helpers.hpp"
+
+using namespace cl::sycl;
+
+constexpr size_t N = 3;
+using Vector = float[N];
+
+Vector a = { 1, 2, 3 };
+
+auto test1 = [](auto program, auto b_queue) {
+
+  kernel k { boost::compute::kernel { program, "zeroify" } };
+
+  queue q { b_queue };
+  cl::sycl::context host_context;
+  cl::sycl::context device_context = q.get_context();
+
+  buffer<int> A { std::begin(a), std::end(a) };
+
+  VERIFY_COND(A.is_data_up_to_date(host_context));
+  VERIFY_COND(!A.is_data_up_to_date(device_context));
+
+  q.submit([&](handler &cgh) {
+      cgh.set_args(A.get_access<access::mode::write>(cgh));
+      cgh.parallel_for(N, k);
+    });
+
+  VERIFY_COND(A.is_data_up_to_date(host_context));
+  VERIFY_COND(!A.is_data_up_to_date(device_context));
+
+  q.wait();
```

Comment:
I do not understand what the code does.
Since the code is asynchronous with a wait in the middle, I guess there is a race condition on the `VERIFY_COND()` before...
I have the feeling that the only observable state is after the `wait()`, not between the submission and the wait.